feat: implement `Base.trylock`, subtype `Base.AbstractLock` for `ReadWriteLock`

AayushSabharwal opened the pull request:
## Summary

- make `ReadWriteLock` subtype `Base.AbstractLock`
- implement writer-side `trylock` with an atomic `0 => -MaxReaders` transition
- make `islocked` report either active readers or a writer
- cover generic `trylock(f, lock)`, `Base.Lockable`, and failed-trylock accounting

This preserves @AayushSabharwal's two original commits and adds the current-main
merge plus a focused correctness follow-up.

## Correctness

The original subtract-and-rollback implementation briefly published a negative
`readercount`. A reader unlocking during that window could decrement
`readerwait`, leaving accounting debt after rollback and allowing a later writer
to overlap a live reader.

The follow-up instead acquires the underlying writer lock and atomically replaces
`readercount` only when it is exactly zero. A failed attempt does not change
reader accounting and does not need rollback.

## Validation

- exact `ReadWriteLock` testset:
  - Julia 1.8.5: 53/53
  - Julia 1.11.9: 54/54
  - Julia 1.12.6: 54/54
  - Julia nightly: 55/55
- AllocArrays 1.1.0 downstream suite on Julia 1.11.9: 461/461
- `git diff --check`

Closes #42.

Co-authored by Codex

AayushSabharwal committed:
refactor: make `ReadWriteLock` subtype `Base.AbstractLock`

AayushSabharwal committed:
feat: implement `Base.trylock` for `ReadWriteLock`

oscardssmith commented on src/rwlock.jl:
```diff
@@ -95,6 +95,23 @@ function Base.lock(rw::ReadWriteLock)
     return
 end
 
+function Base.trylock(rw::ReadWriteLock)
```
is this trying to readlock or write lock?

AayushSabharwal replied:
Writelock, since `lock` and `unlock` are for writelock. The readlock equivalent would likely be a custom `tryreadlock` function.

quinnj committed:
chore(rwlock): merge current main

quinnj committed:
fix(rwlock): make trylock atomic

Use a compare-and-swap to acquire the writer state without publishing transient accounting or requiring rollback.

Report active readers from islocked and cover generic lock APIs, Base.Lockable integration, and failed-trylock accounting.

Fixes #42
